pkg/messageprocessors/normalizedpayload/uplink.go

```diff
@@ -0,0 +1,235 @@
+	"golang.org/x/exp/constraints"
+)
+
+// Air is an air measurement.
+type Air struct {
+	Temperature      *float64
+	RelativeHumidity *float64
+	Pressure         *float64
+}
+
+// Wind is a wind measurement.
+type Wind struct {
+	Speed     *float64
+	Direction *float64
+}
+
+// Measurement is a measurement.
+type Measurement struct {
+	Time *time.Time
+	Air  *Air
+	Wind *Wind
+}
+
+var (
+	errFieldType    = errors.DefineInvalidArgument("field_type", "invalid field type of `{path}`")
+	errFieldMinimum = errors.DefineInvalidArgument(
+		"field_minimum",
+		"`{path}` should be equal or greater than `{minimum}`",
+	)
+	//nolint:unused
+	errFieldExclusiveMinimum = errors.DefineInvalidArgument(
+		"field_exclusive_minimum",
+		"`{path}` should be greater than `{minimum}`",
+	)
+	errFieldMaximum = errors.DefineInvalidArgument(
+		"field_maximum",
+		"`{path}` should be equal or less than `{maximum}`",
+	)
+	errFieldExclusiveMaximum = errors.DefineInvalidArgument(
+		"field_exclusive_maximum",
+		"`{path}` should be less than `{maximum}`",
+	)
+	errUnknownField = errors.DefineInvalidArgument("unknown_field", "unknown field `{path}`")
+)
+
+type fieldParser func(dst *Measurement, src *pbtypes.Value, path string) error
+
+// object validates that the path is a structure and sets the target to an empty value.
+func object[T any](selector func(*Measurement) **T) fieldParser {
+	return func(dst *Measurement, src *pbtypes.Value, path string) error {
+		_, ok := src.Kind.(*pbtypes.Value_StructValue)
+		if !ok {
+			return errFieldType.WithAttributes("path", path)
+		}
+		*selector(dst) = new(T)
+		return nil
+	}
+}
+
+type fieldValidator[T any] func(v T, path string) error
+
+func validate[T any](val T, validators []fieldValidator[T], path string) error {
+	for _, v := range validators {
+		if err := v(val, path); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// parseTime parses and validates the time. The input value must be RFC3339.
+func parseTime(selector func(dst *Measurement) **time.Time, vals ...fieldValidator[time.Time]) fieldParser {
+	return func(dst *Measurement, src *pbtypes.Value, path string) error {
+		val, ok := src.Kind.(*pbtypes.Value_StringValue)
+		if !ok {
+			return errFieldType.WithAttributes("path", path)
+		}
+		t, err := time.Parse(time.RFC3339, val.StringValue)
+		if err != nil {
+			return err
+		}
+		if err := validate(t, vals, path); err != nil {
+			return err
+		}
+		*selector(dst) = &t
+		return nil
+	}
+}
+
+// parseNumber parses and validates a number.
+func parseNumber(selector func(dst *Measurement) **float64, vals ...fieldValidator[float64]) fieldParser {
+	return func(dst *Measurement, src *pbtypes.Value, path string) error {
+		val, ok := src.Kind.(*pbtypes.Value_NumberValue)
+		if !ok {
+			return errFieldType.WithAttributes("path", path)
+		}
+		n := val.NumberValue
+		if err := validate(n, vals, path); err != nil {
+			return err
+		}
+		*selector(dst) = &n
```

Comment:
If the field exists, there's an early exit on a mismatched type. Only if the field name and type are normalized, the actual validations are run. That is not gonna be more than bounds checks, maybe regex, maybe enums.

What we could do, is first check if the structure is normalized, before validating the fields. That would make things a bit faster. In any case, this all happens in locally so I expect the performance impact is negligble. (3/3)